db2util.c

```diff
@@ -2,701 +2,353 @@
+  check_error(hstmt, SQL_HANDLE_STMT, rc);
+
   /* number of input parms */
-  rc = SQLNumParams((SQLHSTMT)hstmt, (SQLSMALLINT*)&nParms);
-  if (db2util_check_sql_errors(fmt, hstmt, SQL_HANDLE_STMT, rc) == SQL_ERROR) {
-    return SQL_ERROR;
+  rc = SQLNumParams(hstmt, &param_count);
+  check_error(hstmt, SQL_HANDLE_STMT, rc);
+
+  if (param_count != argc) {
+    fprintf(stderr, "Invalid parameter count: expected %d got %d\n", param_count, argc);
+    exit(1);
   }
-  if (nParms > 0) {
-    for (i = 0; i < nParms; i++) {
-      rc = SQLDescribeParam((SQLHSTMT)hstmt, (SQLUSMALLINT)(i + 1), 
-             &sql_data_type, &sql_precision, &sql_scale, &sql_nullable);
-      if (db2util_check_sql_errors(fmt, hstmt, SQL_HANDLE_STMT, rc) == SQL_ERROR) {
-        return SQL_ERROR;
-      }
-      buff_len[i] = SQL_NTS;
-      rc = SQLBindParameter((SQLHSTMT)hstmt, (SQLUSMALLINT)(i + 1),
-             SQL_PARAM_INPUT, SQL_CHAR, sql_data_type,
-             sql_precision, sql_scale, argv[i], 0, &buff_len[i]);
-      if (db2util_check_sql_errors(fmt, hstmt, SQL_HANDLE_STMT, rc) == SQL_ERROR) {
-        return SQL_ERROR;
-      }
-    }
+  
+  SQLINTEGER input_indicator = SQL_NTS;
+  for (int i = 0; i < param_count; i++) {
+    SQLSMALLINT type;
+    SQLUINTEGER precision;
+    SQLSMALLINT scale;
+    SQLSMALLINT nullable;
+
+    rc = SQLDescribeParam(hstmt, i+1, &type, &precision, &scale, &nullable);
+    check_error(hstmt, SQL_HANDLE_STMT, rc);
+
+    rc = SQLBindParameter(hstmt, i+1, SQL_PARAM_INPUT, SQL_C_CHAR, type,
+                          precision, scale, (SQLCHAR*) argv[i], 0, &input_indicator);
+    check_error(hstmt, SQL_HANDLE_STMT, rc);
   }
+
   /* execute */
-  rc = SQLExecute((SQLHSTMT)hstmt);
-  if (db2util_check_sql_errors(fmt, hstmt, SQL_HANDLE_STMT, rc) == SQL_ERROR) {
-    return SQL_ERROR;
-  }
+  rc = SQLExecute(hstmt);
+  check_error(hstmt, SQL_HANDLE_STMT, rc);
+
   /* result set */
-  rc = SQLNumResultCols((SQLHSTMT)hstmt, &nResultCols);
-  if (db2util_check_sql_errors(fmt, hstmt, SQL_HANDLE_STMT, rc) == SQL_ERROR) {
-    return SQL_ERROR;
+  rc = SQLNumResultCols(hstmt, &column_count);
+  check_error(hstmt, SQL_HANDLE_STMT, rc);
+
+  if (column_count < 1) exit(0);
+
+  col_info_t* cols = malloc(sizeof(col_info_t) * column_count);
+ 
+  for (int i = 0 ; i < column_count; i++) {
+    col_info_t* col = &cols[i];
+
+    SQLUINTEGER size;
+    SQLSMALLINT ignore;
+
+    rc = SQLDescribeCol(hstmt, i+1, col->name, sizeof(col->name), &ignore, &col->type, &size, &ignore, &ignore);
+    check_error(hstmt, SQL_HANDLE_STMT, rc);
+
+    size_t bind_length;
+    switch (col->type) {
+    case SQL_CHAR:
+    case SQL_VARCHAR:
+    case SQL_CLOB:
+    case SQL_DBCLOB:
+    case SQL_UTF8_CHAR:
+    case SQL_WCHAR:
+    case SQL_WVARCHAR:
+    case SQL_GRAPHIC:
+    case SQL_VARGRAPHIC:
+    case SQL_XML:
+    default:
+      col->bind_type = SQL_C_CHAR;
+      col->buffer_length = bind_length = size * 3;
```

Comment:
What is the idea behind `size * 3`, that it should cover all of the cases? Trying to understand what `size` actually returns, the CLI docs are opaque for me, and the ODBC docs say `return the size (in characters)`, which to me sounds like `(size + 1) * sizeof(SQLCHAR)` or `* sizeof(SQLWCHAR)`, and even then the + 1 just being for the null-terminator.

This is something I have struggled with in ODBC so just curious as to the reasoning here.

Reply:
For ODBC, the relevant documentation can be found [here](https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/column-size-decimal-digits-transfer-octet-length-and-display-size?view=sql-server-2017).

For IBM i CLI, the documentation says "characters", but that's sort of wrong (depending on your understanding of a "character"). Ultimately for character and binary types, IBM i CLI returns the size of the column as it was specified in SQL. eg. `CHAR(100)` -> 100, `VARGRAGPHIC(50)` -> 50.

It's important to remember that this size is the maximum number of "characters" that that column can contain, however how many bytes in UTF-8 that requires to hold is completely unknown until conversion is done. Tony just multiplied things by 2, 3, 4 (and in some cases 6 :astonished:) "just in case".

We could do better heuristics here based on source CCSID, but for now I pretty much left it as is, since it's unlikely to really matter in practice.